### PR TITLE
feat: type alias

### LINKED ISSUE #110: feat: add type aliases
We want to introduce type aliases, as follows:

```rust
type Typename = OtherTypeName;
```

In large, this change builds on existing architecture, but it will require work in several places of our stack:
- [ ] lexer: add the `type` keyword ([src](https://github.com/mun-lang/mun/blob/master/crates/mun_syntax/src/grammar.ron), [example](https://github.com/rust-analyzer/rust-analyzer/blob/363718ba42d1eb906577a0a03c14248a0962a06a/crates/ra_syntax/src/grammar.ron#L90))
- [ ] parsing: add parsing of `type TypeName = OtherTypeName` ([example#1](https://github.com/rust-analyzer/rust-analyzer/blob/363718ba42d1eb906577a0a03c14248a0962a06a/crates/ra_syntax/src/grammar.ron#L139), [example#2](https://github.com/rust-analyzer/rust-analyzer/blob/363718ba42d1eb906577a0a03c14248a0962a06a/crates/ra_parser/src/grammar/items.rs#L336))
- [ ] hir: add type aliases as type definition ([src](https://github.com/mun-lang/mun/tree/master/crates/mun_hir/src))
- [ ] tests: add test for all of the above stages

**Good first issue process:**

- [ ] [read our contributing guidelines](https://github.com/mun-lang/mun/blob/master/CONTRIBUTING.md)
- [ ] claim this issue (assign yourself or comment below)
- [ ] implement the feature and start a Pull Request with in the description: closes #110. 

If this is your first PR, welcome 🎉 😄

### PR DESCRIPTION
**Pull request by sinato**
This PR adds the ability to use type alias in Mun.  (maybe close #110 )

The following code should work properly.

```rust
type Foo = u32;
fn main() {
    let a: Foo = 1;
}
```

And the following code should raise a type mismatch.

```rust
type Foo = u32;
fn main() {
    let a: Foo = 1.0;
}
```
```
error: mismatched type
 --> main.mun:3:17
  |
3 |     let a: Foo = 1.0;
  |                  ^^^ expected `u32`, found `{float}`
  |
```

NOTE: The following code does not work properly yet.
```rust
fn main() {
   type Foo = u32;
   let a: Foo = 10;
}
```

- [x] lexer: add the type keyword
- [x] parsing: add parsing of type TypeName = OtherTypeName
- [x] hir: add type aliases as type definition
- [x] tests: add test for all of the above stages
  - lexer: `crates/mun_syntax/src/tests/snapshots/mun_syntax__tests__lexer__keywords.snap`
  - parser: `crates/mun_syntax/src/tests/snapshots/mun_syntax__tests__parser__type_alias_def.snap`
  - HIR: `crates/mun_hir/src/ty/snapshots/mun_hir__ty__tests__infer_type_alias.snap`

### CODE REVIEW

File: crates/mun_hir/src/expr/validator.rs [hunk 1]
```diff
@@ -83,3 +87,22 @@ impl<'a> ExprValidator<'a> {
...
     }
 }
+
+pub struct TypeAliasValidator<'a> {
+    type_alias: TypeAlias,
+    db: &'a dyn HirDatabase,
+}
+
+impl<'a> TypeAliasValidator<'a> {
+    pub fn new(type_alias: TypeAlias, db: &'a dyn HirDatabase) -> Self {
```
[the remark this answers is not in the document]
Comment (sinato) [reply crates/mun_hir/src/expr/validator.rs:98]: Added simple documentation 7fceca5

**Comment by sinato** [comment]
If there is anything I can (or should) do to help with the stress test, please let me know.

**Comment by Wodann** [comment]
You can try to use `type` in weird ways or unexpected places. E.g. `fn type Foo = u32;` or `(type Foo =u32)` 

If you install the Mun language server for VSCode, you should get continuous feedback to pinpoint errors and crashes more easily.

**Comment by baszalmstra** [comment]
I did a bit of stress testing and found this issue:
```rust
#[test]
fn recursive_alias() {
    infer_snapshot(
        r#"
    struct Foo {}
    type Foo = Foo;

    type A = B
    type B = A;
    "#,
    )
}
```

But cases seem to result in a cycle in the salsa database. 

Also, would you be able create a PR on this repo: https://github.com/mun-lang/vscode-extension to support type aliases in the syntax highlighting?

**Comment by sinato** [comment]
> Both cases seem to result in a cycle in the salsa database.

Thank you for testing! Fixed 00d23e8
I referred to the following commit. https://github.com/rust-analyzer/rust-analyzer/pull/2453/commits/1c622e9fed97de7711da7b5bffec0fa4b19d7500

> Also, would you be able create a PR on this repo: https://github.com/mun-lang/vscode-extension to support type aliases in the syntax highlighting?

Of course. I'll fix it.

**Comment by Wodann** [comment]
Very quick fix; nice! There seems to be a merge conflict, but other than that I am happy to merge this. As far as I am concerned, you can go ahead with the rebase 🙂

**Comment by sinato** [comment]
Rebase completed.

**Comment by baszalmstra** [comment]
Your solution for recursive types is still incomplete. There are no errors reported for the recursive aliases which will crash the codegen. The codegen assumes that if there are no errors the HIR is in a valid state. You can add this test to `crates/mun_codegen/src/test.rs` to witness the problem:

```
#[test]
fn recursive_alias() {
    test_snapshot(
        r#"
    type Alias = Alias;

    fn foo() -> Alias {
    }
        "#,
    )
}
```

I see there is actually [a comment](https://github.com/mun-lang/mun/blob/7147826d99913e052fb0dd0f008174d0d58f92f9/crates/mun_hir/src/ty/lower.rs#L121) that states we need to implement cyclic types. 😅 

This also doesn't throw an error but does break the compiler:

```
#[test]
fn recursive_struct() {
    test_snapshot(
        r#"
    struct(value) Struct {
        a: Struct
    }
        "#,
    )
}
```

@Wodann WDYT we should do? Merge this PR and file an issue? Or should at least the alias cycle thing be fixed?

**Comment by Wodann** [comment]
> @Wodann WDYT we should do? Merge this PR and file an issue? Or should at least the alias cycle thing be fixed?
@sinato do you think you can come up with a fix for the alias cycle based on @baszalmstra's comments or would you like some help with it? As we are aware of the issue before merging, I'd say that we should ideally resolve it first.

**Comment by sinato** [comment]
@Wodann Maybe there are better implementations, but I think the fix would be as 44ae87d .
I'd be happy to get some advice on this fix.

**Review by Wodann — CHANGES_REQUESTED** [review]
Thanks for solving the issue with cyclic type references. This seems like a good enough approach for now.

I found one typo, could you fix that? Then I am happy to merge it 🙂

**Comment by baszalmstra** [comment]
Yes! Fine for me too!

**Comment by sinato** [comment]
Thank you for your comment! Fix typo and squashed the commits.

**Comment by baszalmstra** [comment]
Thank you very much for your contribution and the quick turnaround regarding comments! Great job! 👍👍

**Comment by Wodann** [comment]
Yes, thank you so much, @sinato! I can't wait for your next contribution; it's been a pleasure collaborating. We could really use another regular contributor :slight_smile:

**Comment by sinato** [comment]
It was a very nice experience. Thank you for your kind comments!

**Pull request merged**